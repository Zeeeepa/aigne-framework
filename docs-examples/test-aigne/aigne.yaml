chat_model:
  name: gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
tools:
  - sandbox.js
