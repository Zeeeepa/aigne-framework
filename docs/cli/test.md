# Testing

Use AIGNE CLI to run agent tests.

## `aigne test` Command

Run tests in the specified agent directory.

```bash
aigne test [options]
```

### Options

* `--path <path>`: Path to the agent directory (defaults to current directory `.`)
* `--help`: Show command help

### Examples

#### Run tests in current directory

```bash
aigne test
```

#### Run tests in specific directory

```bash
aigne test --path ./my-agents
```

### Test File Structure

The AIGNE testing system automatically finds and runs test files in your project. Make sure your test files follow the correct naming conventions and structure.

### Best Practices

1. **Organize test files**: Place test files in appropriate directory structures
2. **Write clear test cases**: Ensure each test case has a clear purpose and expected outcome
3. **Use descriptive test names**: Make test names clearly express what the test covers
4. **Run tests regularly**: Run tests regularly during development to ensure code quality
