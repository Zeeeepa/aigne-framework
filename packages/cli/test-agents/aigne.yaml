name: test_aigne_project
description: A test project for the aigne agent
chat_model:
  model: openai:gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
skills:
  - sandbox.js
mcp_server:
  agents:
    - chat.yaml
cli:
  agents:
    - chat.yaml
