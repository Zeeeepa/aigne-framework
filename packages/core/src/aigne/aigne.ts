import { z } from "zod";
import {
  Agent,
  type AgentResponse,
  type AgentResponseStream,
  type Message,
} from "../agents/agent.js";
import type { UserAgent } from "../agents/user-agent.js";
import { load } from "../loader/index.js";
import { ChatModel } from "../models/chat-model.js";
import { checkArguments, createAccessorArray } from "../utils/type-utils.js";
import { AIGNEContext, type InvokeOptions } from "./context.js";
import {
  type MessagePayload,
  MessageQueue,
  type MessageQueueListener,
  type Unsubscribe,
} from "./message-queue.js";
import type { ContextLimits } from "./usage.js";

/**
 * Options for the AIGNE class.
 */
export interface AIGNEOptions {
  /**
   * The name of the AIGNE instance.
   */
  name?: string;

  /**
   * The description of the AIGNE instance.
   */
  description?: string;

  /**
   * Global model to use for all agents not specifying a model.
   */
  model?: ChatModel;

  /**
   * Skills to use for the AIGNE instance.
   */
  skills?: Agent[];

  /**
   * Agents to use for the AIGNE instance.
   */
  agents?: Agent[];

  /**
   * Limits for the AIGNE instance, such as timeout, max tokens, max invocations, etc.
   */
  limits?: ContextLimits;
}

/**
 * AIGNE is a class that orchestrates multiple agents to build complex AI applications.
 * It serves as the central coordination point for agent interactions, message passing, and execution flow.
 *
 * @example
 * Here's a simple example of how to use AIGNE:
 * {@includeCode ../../test/aigne/aigne.test.ts#example-simple}
 *
 * @example
 * Here's an example of how to use AIGNE with streaming response:
 * {@includeCode ../../test/aigne/aigne.test.ts#example-streaming}
 */
export class AIGNE {
  /**
   * Loads an AIGNE instance from a directory containing an aigne.yaml file and agent definitions.
   * This static method provides a convenient way to initialize an AIGNE system from configuration files.
   *
   * @param path - Path to the directory containing the aigne.yaml file.
   * @param options - Options to override the loaded configuration.
   * @returns A fully initialized AIGNE instance with configured agents and skills.
   */
  static async load(path: string, options?: AIGNEOptions): Promise<AIGNE> {
    const { model, agents, skills, ...aigne } = await load({ path });
    return new AIGNE({
      ...options,
      model: options?.model || model,
      name: options?.name || aigne.name || undefined,
      description: options?.description || aigne.description || undefined,
      agents: agents.concat(options?.agents ?? []),
      skills: skills.concat(options?.skills ?? []),
    });
  }

  /**
   * Creates a new AIGNE instance with the specified options.
   *
   * @param options - Configuration options for the AIGNE instance including name, description, model, and agents.
   */
  constructor(options?: AIGNEOptions) {
    if (options) checkArguments("AIGNE", aigneOptionsSchema, options);

    this.name = options?.name;
    this.description = options?.description;
    this.model = options?.model;
    this.limits = options?.limits;
    if (options?.skills?.length) this.skills.push(...options.skills);
    if (options?.agents?.length) this.addAgent(...options.agents);

    this.initProcessExitHandler();
  }

  /**
   * Optional name identifier for this AIGNE instance.
   */
  name?: string;

  /**
   * Optional description of this AIGNE instance's purpose or functionality.
   */
  description?: string;

  /**
   * Global model to use for all agents that don't specify their own model.
   */
  model?: ChatModel;

  /**
   * Usage limits applied to this AIGNE instance's execution.
   */
  limits?: ContextLimits;

  /**
   * Message queue for handling inter-agent communication.
   *
   * @hidden
   */
  readonly messageQueue = new MessageQueue();

  /**
   * Collection of skill agents available to this AIGNE instance.
   * Provides indexed access by skill name.
   */
  readonly skills = createAccessorArray<Agent>([], (arr, name) => arr.find((i) => i.name === name));

  /**
   * Collection of primary agents managed by this AIGNE instance.
   * Provides indexed access by agent name.
   */
  readonly agents = createAccessorArray<Agent>([], (arr, name) => arr.find((i) => i.name === name));

  /**
   * Adds one or more agents to this AIGNE instance.
   * Each agent is attached to this AIGNE instance, allowing it to access the AIGNE's resources.
   *
   * @param agents - One or more Agent instances to add to this AIGNE.
   */
  addAgent(...agents: Agent[]) {
    checkArguments("AIGNE.addAgent", aigneAddAgentArgsSchema, agents);

    for (const agent of agents) {
      this.agents.push(agent);

      agent.attach(this);
    }
  }

  /**
   * Creates a new execution context for this AIGNE instance.
   * Contexts isolate state for different flows or conversations.
   *
   * @returns A new AIGNEContext instance bound to this AIGNE.
   */
  newContext() {
    return new AIGNEContext(this);
  }

  /**
   * Creates a user agent for consistent interactions with a specified agent.
   * This method allows you to create a wrapper around an agent for repeated invocations.
   *
   * @param agent - Target agent to be wrapped for consistent invocation
   * @returns A user agent instance that provides a convenient interface for interacting with the target agent
   *
   * @example
   * Here's an example of how to create a user agent and invoke it consistently:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-user-agent}
   */
  invoke<I extends Message, O extends Message>(agent: Agent<I, O>): UserAgent<I, O>;

  /**
   * Invokes an agent with a message and returns both the output and the active agent.
   * This overload is useful when you need to track which agent was ultimately responsible for generating the response.
   *
   * @param agent - Target agent to invoke
   * @param message - Input message to send to the agent (can be a string or a structured message object)
   * @param options.returnActiveAgent - Must be true to return the final active agent
   * @param options.streaming - Must be false to return a response stream
   * @returns A promise resolving to a tuple containing the agent's response and the final active agent
   */
  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message: I | string,
    options: InvokeOptions & { returnActiveAgent: true; streaming?: false },
  ): Promise<[O, Agent]>;

  /**
   * Invokes an agent with a message and returns both a stream of the response and the active agent.
   * This overload is useful when you need streaming responses while also tracking which agent provided them.
   *
   * @param agent - Target agent to invoke
   * @param message - Input message to send to the agent (can be a string or a structured message object)
   * @param options.returnActiveAgent - Must be true to return the final active agent
   * @param options.streaming - Must be true to return a response stream
   * @returns A promise resolving to a tuple containing the agent's response stream and a promise for the final agent
   */
  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message: I | string,
    options: InvokeOptions & { returnActiveAgent: true; streaming: true },
  ): Promise<[AgentResponseStream<O>, Promise<Agent>]>;

  /**
   * Invokes an agent with a message and returns just the output.
   * This is the standard way to invoke an agent when you only need the response.
   *
   * @param agent - Target agent to invoke
   * @param message - Input message to send to the agent (can be a string or a structured message object)
   * @param options - Optional configuration parameters for the invocation
   * @returns A promise resolving to the agent's complete response
   *
   * @example
   * Here's a simple example of how to invoke an agent:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-simple}
   */
  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message: I | string,
    options?: InvokeOptions & { returnActiveAgent?: false; streaming?: false },
  ): Promise<O>;

  /**
   * Invokes an agent with a message and returns a stream of the response.
   * This allows processing the response incrementally as it's being generated.
   *
   * @param agent - Target agent to invoke
   * @param message - Input message to send to the agent (can be a string or a structured message object)
   * @param options - Configuration with streaming enabled to receive incremental response chunks
   * @returns A promise resolving to a stream of the agent's response that can be consumed incrementally
   *
   * @example
   * Here's an example of how to invoke an agent with streaming response:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-streaming}
   */
  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message: I | string,
    options: InvokeOptions & { returnActiveAgent?: false; streaming: true },
  ): Promise<AgentResponseStream<O>>;

  /**
   * General implementation signature that handles all overload cases.
   * This unified signature supports all the different invocation patterns defined by the overloads.
   *
   * @param agent - Target agent to invoke or wrap
   * @param message - Optional input message to send to the agent
   * @param options - Optional configuration parameters for the invocation
   * @returns Either a UserAgent (when no message provided) or a promise resolving to the agent's response
   * with optional active agent information based on the provided options
   */
  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message?: I | string,
    options?: InvokeOptions,
  ): UserAgent<I, O> | Promise<AgentResponse<O> | [AgentResponse<O>, Agent]>;

  invoke<I extends Message, O extends Message>(
    agent: Agent<I, O>,
    message?: I | string,
    options?: InvokeOptions,
  ): UserAgent<I, O> | Promise<AgentResponse<O> | [AgentResponse<O>, Agent]> {
    return new AIGNEContext(this).invoke(agent, message, options);
  }

  /**
   * Publishes a message to the message queue for inter-agent communication.
   * This method broadcasts a message to all subscribers of the specified topic(s).
   * It creates a new context internally and delegates to the context's publish method.
   *
   * @param topic - The topic or array of topics to publish the message to
   * @param payload - The message payload to be delivered to subscribers
   *
   * @example
   * Here's an example of how to publish a message:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-publish-message}
   */
  publish(topic: string | string[], payload: Omit<MessagePayload, "context"> | Message | string) {
    return new AIGNEContext(this).publish(topic, payload);
  }

  /**
   * Subscribes to receive the next message on a specific topic.
   * This overload returns a Promise that resolves with the next message published to the topic.
   * It's useful for one-time message handling or when using async/await patterns.
   *
   * @param topic - The topic to subscribe to
   * @returns A Promise that resolves with the next message payload published to the specified topic
   *
   * @example
   * Here's an example of how to subscribe to a topic and receive the next message:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-publish-message}
   */
  subscribe(topic: string | string[], listener?: undefined): Promise<MessagePayload>;

  /**
   * Subscribes to messages on a specific topic with a listener callback.
   * This overload registers a listener function that will be called for each message published to the topic.
   * It's useful for continuous message handling or event-driven architectures.
   *
   * @param topic - The topic to subscribe to
   * @param listener - Callback function that will be invoked when messages arrive on the specified topic
   * @returns An Unsubscribe function that can be called to cancel the subscription
   *
   * @example
   * Here's an example of how to subscribe to a topic with a listener:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-subscribe-topic}
   */
  subscribe(topic: string | string[], listener: MessageQueueListener): Unsubscribe;

  /**
   * Generic subscribe signature that handles both Promise and listener patterns.
   * This is the implementation signature that supports both overloaded behaviors.
   *
   * @param topic - The topic to subscribe to
   * @param listener - Optional callback function
   * @returns Either a Promise for the next message or an Unsubscribe function
   */
  subscribe(
    topic: string | string[],
    listener?: MessageQueueListener,
  ): Unsubscribe | Promise<MessagePayload>;
  subscribe(
    topic: string | string[],
    listener?: MessageQueueListener,
  ): Unsubscribe | Promise<MessagePayload> {
    return this.messageQueue.subscribe(topic, listener);
  }

  /**
   * Unsubscribes a listener from a specific topic in the message queue.
   * This method stops a previously registered listener from receiving further messages.
   * It should be called when message processing is complete or when the component is no longer interested
   * in messages published to the specified topic.
   *
   * @param topic - The topic to unsubscribe from
   * @param listener - The listener function that was previously subscribed to the topic
   *
   * @example
   * {@includeCode ../../test/aigne/aigne.test.ts#example-subscribe-topic}
   */
  unsubscribe(topic: string | string[], listener: MessageQueueListener) {
    this.messageQueue.unsubscribe(topic, listener);
  }

  /**
   * Gracefully shuts down the AIGNE instance and all its agents and skills.
   * This ensures proper cleanup of resources before termination.
   *
   * @returns A promise that resolves when shutdown is complete.
   *
   * @example
   * Here's an example of shutdown an AIGNE instance:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-shutdown}
   */
  async shutdown() {
    for (const tool of this.skills) {
      await tool.shutdown();
    }
    for (const agent of this.agents) {
      await agent.shutdown();
    }
  }

  /**
   * Asynchronous dispose method for the AIGNE instance.
   *
   * @example
   * Here's an example of using async dispose:
   * {@includeCode ../../test/aigne/aigne.test.ts#example-shutdown-using}
   */
  async [Symbol.asyncDispose]() {
    await this.shutdown();
  }

  /**
   * Initializes handlers for process exit events to ensure clean shutdown.
   * This registers handlers for SIGINT and exit events to properly terminate all agents.
   */
  private initProcessExitHandler() {
    const shutdownAndExit = () => this.shutdown().finally(() => process.exit(0));
    process.on("SIGINT", shutdownAndExit);
    process.on("exit", shutdownAndExit);
  }
}

const aigneOptionsSchema = z.object({
  model: z.instanceof(ChatModel).optional(),
  skills: z.array(z.instanceof(Agent)).optional(),
  agents: z.array(z.instanceof(Agent)).optional(),
});

const aigneAddAgentArgsSchema = z.array(z.instanceof(Agent));
