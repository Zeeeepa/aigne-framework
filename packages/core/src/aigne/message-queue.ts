import { EventEmitter } from "node:events";
import { z } from "zod";
import type { Message } from "../agents/agent.js";
import { createMessage } from "../prompt/prompt-builder.js";
import { checkArguments, isNil, orArrayToArray } from "../utils/type-utils.js";
import type { Context } from "./context.js";

/**
 * @hidden
 */
export const UserInputTopic = "UserInputTopic";

/**
 * @hidden
 */
export const UserOutputTopic = "UserOutputTopic";

/**
 * @hidden
 */
export interface MessagePayload {
  role: "user" | "agent";
  source?: string;
  message: Message;

  // Context of the message
  context: Context;
}

function isMessagePayload(payload: unknown): payload is Omit<MessagePayload, "context"> {
  return (
    !isNil(payload) &&
    typeof payload === "object" &&
    "role" in payload &&
    typeof payload.role === "string" &&
    ["user", "agent"].includes(payload.role) &&
    "message" in payload &&
    !isNil(payload.message)
  );
}

/**
 * @hidden
 */
export function toMessagePayload(
  payload: Omit<MessagePayload, "context"> | string | Message,
  options?: Partial<Pick<MessagePayload, "role" | "source">>,
): Omit<MessagePayload, "context"> {
  if (isMessagePayload(payload)) {
    return { ...payload, message: createMessage(payload.message), ...options };
  }
  return {
    role: options?.role || "user",
    source: options?.source,
    message: createMessage(payload),
  };
}

/**
 * @hidden
 */
export type MessageQueueListener = (message: MessagePayload) => void;

/**
 * @hidden
 */
export type Unsubscribe = () => void;

/**
 * @hidden
 */
export class MessageQueue {
  events = new EventEmitter();

  publish(topic: string | string[], payload: MessagePayload) {
    checkArguments("MessageQueue.publish", publishArgsSchema, {
      topic,
      payload,
    });

    for (const t of orArrayToArray(topic)) {
      this.events.emit(t, payload);
    }
  }

  error(error: Error) {
    this.events.emit("error", error);
  }

  subscribe(topic: string | string[], listener?: undefined): Promise<MessagePayload>;
  subscribe(topic: string | string[], listener: MessageQueueListener): Unsubscribe;
  subscribe(
    topic: string | string[],
    listener?: MessageQueueListener,
  ): Unsubscribe | Promise<MessagePayload>;
  subscribe(
    topic: string | string[],
    listener?: MessageQueueListener,
  ): Unsubscribe | Promise<MessagePayload> {
    checkArguments("MessageQueue.subscribe", subscribeArgsSchema, {
      topic,
      listener,
    });

    if (!listener) {
      return new Promise((resolve, reject) => {
        const unsubscribe1 = once<MessagePayload>(this.events, topic, (message) => {
          unsubscribe2();
          resolve(message);
        });
        const unsubscribe2 = once(this.events, "error", (error) => {
          unsubscribe1();
          reject(error);
        });
      });
    }

    return on(this.events, topic, listener);
  }

  unsubscribe(topic: string | string[], listener: MessageQueueListener) {
    checkArguments("MessageQueue.unsubscribe", unsubscribeArgsSchema, {
      topic,
      listener,
    });

    for (const t of orArrayToArray(topic)) {
      this.events.off(t, listener);
    }
  }
}

function on<T>(
  events: EventEmitter,
  event: string | string[],
  listener: (arg: T, ...args: unknown[]) => void,
): Unsubscribe {
  orArrayToArray(event).forEach((e) => events.on(e, listener));
  return () => orArrayToArray(event).forEach((e) => events.off(e, listener));
}

function once<T>(
  events: EventEmitter,
  event: string | string[],
  listener: (arg: T, ...args: unknown[]) => void,
): Unsubscribe {
  orArrayToArray(event).forEach((e) => events.once(e, listener));
  return () => orArrayToArray(event).forEach((e) => events.off(e, listener));
}

const subscribeArgsSchema = z.object({
  topic: z.union([z.string(), z.array(z.string())]),
  listener: z.function(z.tuple([z.any()]), z.any()).optional(),
});

const unsubscribeArgsSchema = z.object({
  topic: z.union([z.string(), z.array(z.string())]),
  listener: z.function(z.tuple([z.any()]), z.any()),
});

const publishArgsSchema = z.object({
  topic: z.union([z.string(), z.array(z.string())]),
  payload: z.object({
    role: z.union([z.literal("user"), z.literal("agent")]),
    source: z.string().optional(),
    message: z.union([z.string(), z.record(z.unknown())]),
    context: z.any(),
  }),
});
