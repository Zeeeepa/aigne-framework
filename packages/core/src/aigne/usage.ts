/**
 * @hidden
 */
export interface ContextUsage {
  inputTokens: number;
  outputTokens: number;
  aigneHubCredits: number;
  agentCalls: number;
  duration: number;
}

/**
 * @hidden
 */
export function newEmptyContextUsage(): ContextUsage {
  return {
    inputTokens: 0,
    outputTokens: 0,
    aigneHubCredits: 0,
    agentCalls: 0,
    duration: 0,
  };
}

/**
 * @hidden
 */
export interface ContextLimits {
  maxTokens?: number;
  maxAgentInvokes?: number;
  timeout?: number;
}

export function mergeContextUsage(usage: ContextUsage, additional: Partial<ContextUsage>) {
  if (additional.inputTokens) usage.inputTokens += additional.inputTokens;
  if (additional.outputTokens) usage.outputTokens += additional.outputTokens;
  if (additional.aigneHubCredits) usage.aigneHubCredits += additional.aigneHubCredits;
  if (additional.agentCalls) usage.agentCalls += additional.agentCalls;
  if (additional.duration) usage.duration += additional.duration;
}
