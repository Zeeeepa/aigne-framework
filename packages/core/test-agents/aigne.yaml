name: test_aigne_project
description: A test project for the aigne agent
chat_model:
  name: gpt-4o-mini
  temperature: 0.8
agents:
  - chat.yaml
  - chat-with-prompt.yaml
  - team.yaml
  - image.yaml
skills:
  - sandbox.js
mcp_server:
  agents:
    - chat.yaml
cli:
  agents:
    - chat.yaml
