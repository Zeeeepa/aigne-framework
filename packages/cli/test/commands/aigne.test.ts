import { expect, spyOn, test } from "bun:test";
import { createAIGNECommand } from "@aigne/cli/commands/aigne.js";

test("aigne command should parse --version correctly", async () => {
  const command = createAIGNECommand();

  spyOn(process, "exit").mockReturnValueOnce(0 as never);

  const result = await command.parse(["--version"]);

  expect(result).toEqual(
    expect.objectContaining({
      version: true,
    }),
  );
});

test("aigne command should print help if no any subcommand", async () => {
  const command = createAIGNECommand();

  const exit = spyOn(process, "exit").mockReturnValue(undefined as never);
  spyOn(console, "error")
    .mockReturnValue(undefined as never)
    .mockRestore();
  const log = spyOn(console, "error").mockReturnValue(undefined as never);

  await command.parseAsync([]);

  expect(log.mock.calls).toMatchSnapshot();

  exit.mockRestore();
  log.mockRestore();
});
