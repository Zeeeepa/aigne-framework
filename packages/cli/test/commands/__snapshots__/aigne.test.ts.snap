// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`aigne command should print help if no any subcommand 1`] = `
[
  [
    
"CLI for AIGNE framework

Commands:
  aigne run [path]     Run AIGNE from the specified agent
  aigne test           Run tests in the specified agents directory
  aigne create [path]  Create a new aigne project with agent config files
  aigne serve-mcp      Serve the agents in the specified directory as a MCP
                       server (streamable http)
  aigne observe        Start the observability server
  aigne doc-smith      Generate and maintain project docs â€” powered by agents.
                                                        [aliases: docsmith, doc]
  aigne hub <command>  Manage AIGNE Hub connections
  aigne deploy         Deploy an aigne application

Options:
  -h, --help     Show help                                             [boolean]
  -v, --version  Show version number                                   [boolean]"
,
  ],
  [],
  [
    "Not enough non-option arguments: got 0, need at least 1",
  ],
]
`;
